fiftyone/core/odm/database.py

```diff
@@ -795,6 +795,172 @@ def list_datasets():
+
+    Args:
+        dataset_name: the name of the dataset
+        dry_run (False): whether to log the actions that would be taken but not
+            perform them
+    """
+    _patch_runs(
+        dataset_name,
+        "annotation_runs",
+        "annotation run",
+        dry_run=dry_run,
+    )
+
+
+def patch_brain_runs(dataset_name, dry_run=False):
+    """Ensures that the brain method runs in the ``runs`` collection for the
+    given dataset exactly match the values in its dataset document.
+
+    Args:
+        dataset_name: the name of the dataset
+        dry_run (False): whether to log the actions that would be taken but not
+            perform them
+    """
+    _patch_runs(
+        dataset_name,
+        "brain_methods",
+        "brain method run",
+        dry_run=dry_run,
+    )
+
+
+def patch_evaluations(dataset_name, dry_run=False):
+    """Ensures that the evaluation runs in the ``runs`` collection for the
+    given dataset exactly match the values in its dataset document.
+
+    Args:
+        dataset_name: the name of the dataset
+        dry_run (False): whether to log the actions that would be taken but not
+            perform them
+    """
+    _patch_runs(dataset_name, "evaluations", "evaluation", dry_run=dry_run)
+
+
+def _patch_runs(dataset_name, runs_field, run_str, dry_run=False):
+    conn = get_db_conn()
+    _logger = _get_logger(dry_run=dry_run)
+
+    dataset_dict = conn.datasets.find_one({"name": dataset_name})
+    if not dataset_dict:
+        _logger.warning("Dataset '%s' not found", dataset_name)
+        return
+
+    dataset_id = dataset_dict["_id"]
+
+    # {key: id} in dataset_dict
+    runs_dict = dataset_dict.get(runs_field, {})
+
+    # {key: id} in runs collection
+    rd = {}
+    for run_dict in conn.runs.find({"_dataset_id": dataset_id}):
+        try:
+            rd[run_dict["key"]] = run_dict["_id"]
+        except:
+            pass
+
+    # In order for a run to be valid, its ``(key, id)`` must match in both
+    # `dataset_dict` and the `runs` collection
+    runs = set(runs_dict.items())
+    run_docs = set(rd.items())
+    made_changes = False
+
+    bad_runs = runs - run_docs
+    num_bad_runs = len(bad_runs)
+    if num_bad_runs > 0:
+        _logger.info(
+            "Purging %d %s(s) %s from dataset",
+            num_bad_runs,
+            run_str,
+            bad_runs,
+        )
+        for bad_key, _ in bad_runs:
+            runs_dict.pop(bad_key, None)
+
+        made_changes = True
+
+    missing_runs = run_docs - runs
+    num_missing_runs = len(missing_runs)
+    if num_missing_runs > 0:
+        _logger.info(
+            "Adding %d misplaced %s(s) %s to dataset",
+            num_missing_runs,
+            run_str,
+            missing_runs,
+        )
+        runs_dict.update(missing_runs)
+        made_changes = True
+
+    if made_changes:
+        dataset_dict[runs_field] = runs_dict
+        conn.datasets.replace_one({"name": dataset_name}, dataset_dict)
```

Comment:
since we are updating only a single field, why are we replacing the entire doc?

Reply:
Good call, will switch to `update_one()`